core/src/main/java/org/elasticsearch/common/io/stream/NamedWriteable.java

```diff
@@ -0,0 +1,33 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.common.io.stream;
+
+/**
+ * A {@link Writeable} object identified by its name.
+ * To be used for arbitrary serializable objects (e.g. queries); when reading them, their name tells
+ * which specific object needs to be created.
+ */
+public interface NamedWriteable<T> extends Writeable<T> {
+
+    /**
+     * Returns the name of the writeable object
+     */
+    String getWriteableName();
```

Comment:
it is getName on the query-refactoring branch, but then classes that implement this interface might already have a getName method (e.g. aggs), which is why I went for this more specific method name.